File: Server/scripts/cell/ConfigObject/Skill/SkillBase/RequireDefine.py
# -*- coding: utf-8 -*-


"""
usage:
	技能消耗类

attribute:
	pass

member function:
	pass

callback:
	pass

"""


#common
import csdefine
#data
import csstatus


class Require:
	"""
	基类
	"""
	def __init__( self ):
		"""
		"""
		pass

	def load( self, args ):
		"""
		以字符串作为参数加载；

		@param args: 各字符串具体意义由各派生类自己定义
		@type  args: STRING
		"""
		pass

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。

		@param caster: 施法者
		@type  caster: Entity
		@param skillInstance: 使用的技能实例
		@type  skillInstance: Skill
		"""
		return csstatus.SKILL_GO_ON

	def pay( self, caster, skillInstance ):
		"""
		支付消耗物

		@param caster: 施法者
		@type  caster: Entity
		@param skillInstance: 使用的技能实例
		@type  skillInstance: Skill
		"""
		pass

class RequireMp( Require ):
	"""
	蓝量(MP)消耗
	"""
	def __init__( self ):
		"""
		构造函数。
		"""
		Require.__init__( self )
		self.value = 0
		self._percent = 0.0

	def load( self, args ):
		"""
		以字符串作为参数加载；

		@param args: format: int
		@type  args: STRING
		"""
		if args["Param1"]:
			self.value = int( args["Param1"] )
		if args["Param2"]:
			self._percent = float( args["Param2"] )

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。
		"""
		if caster.MP >= self.getPay( caster, skillInstance ):
			return csstatus.SKILL_GO_ON
			
		return csstatus.SKILL_OUTOF_MP
		
	def pay( self, caster, skillInstance ):
		"""
		支付消耗物
		"""
		if not self.getPay( caster, skillInstance ): #如果消耗为0则不用去支付消耗
			return
		mp = caster.MP - self.getPay( caster, skillInstance )
		caster.setMP( mp )

	def getPay( self, caster, skillInstance ):
		"""
		获取消耗物
		"""
		_value = int(caster.MP_Max * self._percent) + self.value
		if _value < 0:			# 当计算的消耗小于0时，当作0处理
			_value = 0
		return _value

class RequireHp( Require ):
	"""
	血量(HP)消耗
	"""
	def __init__( self ):
		"""
		构造函数。
		"""
		Require.__init__( self )
		self.value = 0
		self._percent = 0.0

	def load( self, args ):
		"""
		以字符串作为参数加载；

		@param args: format: int
		@type  args: STRING
		"""
		if args["Param1"]:
			self.value = int( args["Param1"] )
		if args["Param2"]:
			self._percent = float( args["Param2"] )

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。
		"""
		if caster.HP > self.getPay( caster, skillInstance ):
			return csstatus.SKILL_GO_ON

		return csstatus.SKILL_OUTOF_HP

	def pay( self, caster, skillInstance ):
		"""
		支付消耗物
		"""
		if not self.getPay( caster, skillInstance ): #如果消耗为0则不用去支付消耗
			return
		hp = caster.HP - self.getPay( caster, skillInstance )
		caster.setHP( hp )

	def getPay( self, caster, skillInstance ):
		"""
		获取消耗物
		"""
		_value = int(caster.HP_Max * self._percent) + self.value
		if _value < 0:
			_value = 0
		return _value

class RequireGangQi( Require ):
	"""
	罡气消耗
	"""
	def __init__( self ):
		"""
		构造函数。
		"""
		Require.__init__( self )
		self.value = 0
		self._percent = 0.0

	def load( self, args ):
		"""
		以字符串作为参数加载；

		@param args: format: int
		@type  args: STRING
		"""
		if args["Param1"]:
			self.value = int( args["Param1"] )
		if args["Param2"]:
			self._percent = float( args["Param2"] )

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。
		"""
		if caster.gangQiValue > self.getPay( caster, skillInstance ):
			return csstatus.SKILL_GO_ON

		return csstatus.SKILL_OUTOF_GANGQI

	def pay( self, caster, skillInstance ):
		"""
		支付消耗物
		"""
		if not self.getPay( caster, skillInstance ):  #如果消耗为0则不用去支付消耗
			return
		_gangQiValue = self.getPay( caster, skillInstance )
		caster.recvGangQiDamage( _gangQiValue )

	def getPay( self, caster, skillInstance ):
		"""
		获取消耗物
		"""
		_value = int(caster.gangQiValue_Max * self._percent) + self.value
		if _value < 0:
			_value = 0
		return _value




class RequireDefine( Require ):

	def __init__( self ):
		"""
		构造函数。
		"""
		Require.__init__( self )
		self._requires = []

	def load( self, pyDat ):
		"""
		以字符串作为参数加载；

		@param dictDat: python data
		"""
		for dat in pyDat:
			instance = eval( dat[ "Type" ] )()
			instance.load( dat )
			self._requires.append( instance )

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。

		注：如果出现类似于“使用技能时，有$1%的几率‘增加’$2%的法力消耗”时，表现不一定正确，
		因为现在没有办法正确表现随机的“增加”
		"""
		for r in self._requires:
			state = r.validObject( caster, skillInstance )
			if state != csstatus.SKILL_GO_ON:
				return state
		return csstatus.SKILL_GO_ON

	def pay( self, caster, skillInstance ):
		"""
		支付消耗物
		"""
		for r in self._requires:
			r.pay( caster, skillInstance )

class RequireCostItem( Require ):
	"""
	物品消耗
	"""
	def __init__( self ):
		"""
		构造函数。
		"""
		Require.__init__( self )
		self.itemID = 0
		self.amount = 0

	def load( self, args ):
		"""
		以字符串作为参数加载；

		@param args: format: int
		@type  args: STRING
		"""
		if args["Param1"]:
			self.itemID = int( args["Param1"] )
		if args["Param2"]:
			self.amount = int( args["Param2"] )

	def validObject( self, caster, skillInstance ):
		"""
		校验对象是否满足需求。
		"""
		itemList = caster.getItemInstsByID( self.itemID )
		if not len(itemList):
			return csstatus.SKILL_OUTOF_ITEM
		totalAmount = 0
		for item in itemList:
			totalAmount += item.amount
		if not (totalAmount >= self.amount):
			return csstatus.SKILL_OUTOF_ITEM
		return csstatus.SKILL_GO_ON

	def pay( self, caster, skillInstance ):
		"""
		支付消耗物
		"""
		itemList = caster.getItemInstsByID( self.itemID )
		caster.removeItemListByAmount(itemList, self.amount, csdefine.ITEM_REMOVE_BY_SKILL)

def newInstance( dictDat ):
	"""
	获取需求实例。
		以字符串作为参数加载；

		@param dictDat: python data
	"""
	inst = RequireDefine()
	if dictDat:
		inst.load( dictDat )
	return inst

File: Client/Chuangshi/Skill/SkillBase/RequireDefine.cpp

#include "RequireDefine.h"
#include "Util/ConvertUtil.h"
#include "Util/GolbalBPFunctionLibrary.h"
#include "Interface/Monster/CSkillInterface.h"
#include "Interface/Role/RoleKitBagInterface.h"
#include "GameData/GameDeFine.h"
#include "Skill.h"

URequire::URequire()
{

}

URequire::~URequire()
{

}

void URequire::load(FSKILL_REQUIRE Args)
{

}

int32 URequire::validObject(CS3Entity* caster, USkill* SkillInstance)
{
	return (int32)SKILL_MESSAGE::SKILL_GO_ON;
}

void URequire::pay(CS3Entity* caster, USkill* SkillInstance)
{

}

int32 URequire::getPay(CS3Entity* caster, USkill* skillInstance)
{
	return 0;
}

URequireMp::URequireMp()
{
}

URequireMp::~URequireMp()
{
}

void URequireMp::load(FSKILL_REQUIRE Args)
{
	if (Args.Param1 != "")
	{
		_value = FSTRING_TO_INT(Args.Param1);
	}
	if (Args.Param2 != "")
	{
		_percent = FSTRING_TO_FLOAT(Args.Param2);
	}
}

int32 URequireMp::validObject(CS3Entity * caster, USkill * SkillInstance)
{
	int32 MP = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("MP"));
	if (MP >= getPay(caster, SkillInstance))
	{
		return (int32)SKILL_MESSAGE::SKILL_GO_ON;
	}
	return (int32)SKILL_MESSAGE::SKILL_OUTOF_MP;
}

void URequireMp::pay(CS3Entity * caster, USkill * SkillInstance)
{
	int value = getPay(caster, SkillInstance);
	if (value == 0)
	{
		return;
	}
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), caster->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->CSkillSetMP(value);
	}
}

int32 URequireMp::getPay(CS3Entity * caster, USkill * skillInstance)
{
	int32 MP_Max = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("MP_Max"));
	int32 RequireValue = _value + int(MP_Max * _percent);
	return RequireValue > 0 ? RequireValue : 0 ;
}

URequireHp::URequireHp()
{
}

URequireHp::~URequireHp()
{
}

void URequireHp::load(FSKILL_REQUIRE Args)
{
	if (Args.Param1 != "")
	{
		_value = FSTRING_TO_INT(Args.Param1);
	}
	if (Args.Param2 != "")
	{
		_percent = FSTRING_TO_FLOAT(Args.Param2);
	}
}

int32 URequireHp::validObject(CS3Entity * caster, USkill * SkillInstance)
{
	int32 HP = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("HP"));
	if (HP >= getPay(caster, SkillInstance))
	{
		return (int32)SKILL_MESSAGE::SKILL_GO_ON;
	}
	return (int32)SKILL_MESSAGE::SKILL_OUTOF_HP;
}

void URequireHp::pay(CS3Entity * caster, USkill * SkillInstance)
{
	int value = getPay(caster, SkillInstance);
	if (value == 0)
	{
		return;
	}
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), caster->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->CSkillSetHP(value);
	}
}

int32 URequireHp::getPay(CS3Entity * caster, USkill * skillInstance)
{
	int32 HP_Max = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("HP_Max"));
	int32 RequireValue = _value + int(HP_Max * _percent);
	return RequireValue > 0 ? RequireValue : 0;
}

URequireGangQi::URequireGangQi()
{
}

URequireGangQi::~URequireGangQi()
{
}

inline void URequireGangQi::load(FSKILL_REQUIRE Args)
{
	if (Args.Param1 != "")
	{
		_value = FSTRING_TO_INT(Args.Param1);
	}
	if (Args.Param2 != "")
	{
		_percent = FSTRING_TO_FLOAT(Args.Param2);
	}
}

inline int32 URequireGangQi::validObject(CS3Entity * caster, USkill * SkillInstance)
{
	int32 gangQiValue = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("gangQiValue"));
	if (gangQiValue >= getPay(caster, SkillInstance))
	{
		return (int32)SKILL_MESSAGE::SKILL_GO_ON;
	}
	return (int32)SKILL_MESSAGE::SKILL_OUTOF_HP;
}

void URequireGangQi::pay(CS3Entity * caster, USkill * SkillInstance)
{
	int value = getPay(caster, SkillInstance);
	if (value == 0)
	{
		return;
	}
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), caster->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->CSkillRecvGangQi(value);
	}
}

int32 URequireGangQi::getPay(CS3Entity * caster, USkill * skillInstance)
{
	int32 gangQiValue_Max = UGolbalBPFunctionLibrary::GetIntPropertyValue(caster->ID(), TEXT("gangQiValue_Max"));
	int32 RequireValue = _value + int(gangQiValue_Max * _percent);
	return RequireValue > 0 ? RequireValue : 0;
}

URequireCostItem::URequireCostItem()
{

}

URequireCostItem::~URequireCostItem()
{
}

void URequireCostItem::load(FSKILL_REQUIRE Args)
{
	if (Args.Param1 != "")
	{
		_itemID = FSTRING_TO_INT(Args.Param1);
	}
	if (Args.Param2 != "")
	{
		_amount = FSTRING_TO_INT(Args.Param2);
	}
}

int32 URequireCostItem::validObject(CS3Entity * caster, USkill * SkillInstance)
{
	URoleKitBagInterface* RoleKitBagInterface = Cast<URoleKitBagInterface>
		(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("RoleKitBagInterface")));

	TArray<UItemBase*> Items = RoleKitBagInterface->GetItemsByItemID(_itemID);
	if (Items.Num() <= 0)
	{
		return (int32)SKILL_MESSAGE::SKILL_OUTOF_ITEM;
	}
	int32 totalAmount = 0;
	for (auto Item : Items)
	{
		totalAmount += Item->Amount;
	}
	if (totalAmount > _amount)
	{
		return (int32)SKILL_MESSAGE::SKILL_OUTOF_ITEM;
	}
	return (int32)SKILL_MESSAGE::SKILL_GO_ON;
}

void URequireCostItem::pay(CS3Entity * caster, USkill * SkillInstance)
{
	UCSkillInterface *CSkillInterface = Cast<UCSkillInterface>(UGolbalBPFunctionLibrary::GetInterfaceByName(TEXT("CSkillInterface"), caster->ID()));
	if (IsValid(CSkillInterface))
	{
		CSkillInterface->CSkillRequireItem(_itemID, _amount);
	}
}

int32 URequireCostItem::getPay(CS3Entity * caster, USkill * skillInstance)
{
	return 0;
}

URequireDefine::URequireDefine()
{

}

URequireDefine::~URequireDefine()
{

}

void URequireDefine::load(TArray<FSKILL_REQUIRE> Args)
{
	for (auto i : Args)
	{
		if (i.Type == ENUM_SKILL_REQUIRE_TYPE::RequireMp)
		{
			URequireMp * Instance = NewObject <URequireMp>();
			Instance->load(i);
			_requires.Add((URequire* )Instance);
		}
		else if (i.Type == ENUM_SKILL_REQUIRE_TYPE::RequireHp)
		{
			URequireHp * Instance = NewObject <URequireHp>();
			Instance->load(i);
			_requires.Add((URequire*)Instance);
		}
		else if (i.Type == ENUM_SKILL_REQUIRE_TYPE::RequireGangQi)
		{
			URequireGangQi * Instance = NewObject <URequireGangQi>();
			Instance->load(i);
			_requires.Add((URequire*)Instance);
		}
		else if (i.Type == ENUM_SKILL_REQUIRE_TYPE::RequireCostItem)
		{
			URequireCostItem * Instance = NewObject <URequireCostItem>();
			Instance->load(i);
			_requires.Add((URequire*)Instance);
		}
		else
		{
			URequire * Instance = NewObject <URequire>();
			Instance->load(i);
			_requires.Add(Instance);
		}
	}
}

int32 URequireDefine::validObject(CS3Entity* caster, USkill* SkillInstance)
{
	for (auto requires : _requires)
	{
		int32 state = requires->validObject(caster, SkillInstance);
		if (state != (int32)SKILL_MESSAGE::SKILL_GO_ON )
		{
			return state;
		}
	}
	return (int32)SKILL_MESSAGE::SKILL_GO_ON;
}

void URequireDefine::pay(CS3Entity* caster, USkill* SkillInstance)
{
	for (auto requires : _requires)
	{
		requires->pay(caster, SkillInstance);
	}
}

